internal/cli/certificate.go

```diff
@@ -0,0 +1,434 @@
+package cli
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	apiclient "github.com/flightctl/flightctl/internal/api/client"
+	"github.com/flightctl/flightctl/internal/client"
+	fccrypto "github.com/flightctl/flightctl/internal/crypto"
+	"github.com/flightctl/flightctl/internal/util/validation"
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"golang.org/x/term"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/yaml"
+)
+
+const secondsInDay = 86400
+const agentPath = "certs/"
+
+type CertificateOptions struct {
+	GlobalOptions
+	Name         string
+	Expiration   string
+	OutputFormat string
+	OutputDir    string
+	EncryptKey   bool
+	SignerName   string
+}
+
+func DefaultCertificateOptions() *CertificateOptions {
+	return &CertificateOptions{
+		GlobalOptions: DefaultGlobalOptions(),
+		Expiration:    "7d",
+		OutputFormat:  "reference",
+		OutputDir:     "./",
+		EncryptKey:    false,
+		SignerName:    "enrollment",
+	}
+}
+
+func NewCmdCertificate() *cobra.Command {
+	o := DefaultCertificateOptions()
+	cmd := &cobra.Command{
+		Use: "certificate request [flags]",
+		// more subcommands may be added later
+		Short:     "request a new enrollment certificate for a device with 'certificate request'",
+		Args:      cobra.MatchAll(cobra.MinimumNArgs(1), cobra.OnlyValidArgs),
+		ValidArgs: []string{"request"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(cmd, args); err != nil {
+				return err
+			}
+			if err := o.Validate(args); err != nil {
+				return err
+			}
+			return o.Run(cmd.Context(), args)
+		},
+		SilenceUsage: true,
+	}
+	o.Bind(cmd.Flags())
+	return cmd
+}
+
+func (o *CertificateOptions) Bind(fs *pflag.FlagSet) {
+	o.GlobalOptions.Bind(fs)
+	fs.StringVarP(&o.Name, "name", "n", "csr", "specify a name for the certificate signing request")
+	fs.StringVarP(&o.Expiration, "expiration", "x", "7d", "specify desired certificate expiration in days, example: 7d")
+	fs.StringVarP(&o.OutputFormat, "output-format", "o", "default", "specify desired output format for an enrollment cert: either 'reference' to have the config file reference key and cert file paths, or 'embedded' to have the key and cert embedded in the config file")
+	fs.StringVarP(&o.OutputDir, "output-dir", "d", "./", "specify desired output directory for key, cert, and ca files (defaults to current directory)")
+	fs.StringVarP(&o.SignerName, "cert-type", "t", "enrollment", "specify the type of certificate requested: 'enrollment' or 'ca' (default 'enrollment')")
```

Comment:
I understand it's the signer not the "type", 

I find it weird that we could ask the CA for a CA certificate.

I guess we are asking for a specific signer.

We need to iron this out, but I am good to do this as a follow up, the functionality you are building here would be useful already there: https://github.com/flightctl/flightctl/pull/569/files#diff-c3579ee33fafd5cc39c2bcc363f2867cf84a172ec353e7d53265439170111719

Reply:
Yeah, it's about having a certificate signed for a specific purpose by a signer specialized on that purpose. Right now, we only have CSRs for enrollment certs, at some point _possibly_ also for the management certs (for cert rotation) or applications.
I believe the "ca" type in here is a leftover from refactoring that we should remove.

Reply:
Yes, we decide/indicate the "type" (enrollment, management, other types) based on the signer, so "type" and "signer" are effectively the same thing.

"ca" is only left in as a type/signer because that is the only other type/signer we have other than "enrollment" and I needed something from which to differentiate "enrollment" in the logic to do auto-approves.